src/synapse-plugins/converter-plugin/converter-plugin-backend.vala

```diff
@@ -0,0 +1,372 @@
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Michal Hruby <michal.mhr@gmail.com>
+*/
+
+namespace Synapse {
+    struct UnitMatch {
+        Unit unit; // Unit that matches in UNITS
+        SIPrefix prefix; // Prefix taken into account
+        int dimension; // Dimension taken into account
+
+        public string description () {
+            string dim = "";
+            if (dimension == 2) {
+                dim = _("squared");
+            } else if (dimension == 3) {
+                dim = _("cubed");
+            }
+
+            /// TRANSLATORS First %s SI prefix, Second %s unit name, Third %s dimension (blank, squared or cubed);
+            return _("%s%s %s").printf (prefix.prefix, unit.description, dim);
+        }
+
+        public double factor () { // Taking into account size, prefix and dimension
+            double factor = 1.0;
+            double size = unit.get_factor ();
+            for (int i = 0; i < dimension; i++) {
+                factor *= prefix.factor;
+                factor *= size;
+            }
+
+            return factor;
+        }
+    }
+
+    public struct ResultData {
+        double factor;
+        string from_description;
+        string to_description;
+    }
+
+    public class ConverterPluginBackend : Object {
+        public static ConverterPluginBackend get_instance () {
+            if (instance == null) {
+                instance = new ConverterPluginBackend ();
+            }
+
+            return instance;
+        }
+
+        private Regex? convert_regex = null;
+        private static ConverterPluginBackend instance = null;
+
+        construct {
+            /* The regex describes a string which *resembles* a unit conversion request in the form
+             * <number> <unit> => <unit>.
+             * Some restrictions are placed on the form of <unit> (letters maybe followed by a 2 or 3).
+             */
+
+            try {
+                convert_regex = new Regex (
+                    """\d*[a-zA-Z\/ ]+[23]?=>[a-zA-Z\/ ]+[23]?""",
+                    RegexCompileFlags.OPTIMIZE
+                );
+            } catch (Error e) {
+                critical ("Error creating regexp: %s", e.message);
+            }
+        }
+
+        public ResultData[] get_conversion_data (string query_string) {
+            ResultData [] results = {};
+            if (!query_string.contains ("=>")) {
+                return results;
+            }
+
+            var input = query_string.replace (",", ".").replace ("|", "").strip ();
```

Comment:
Its probably unnecessary as an entry with that character probably isn't a valid conversion request anyway.  It was probably because that character is used to separate alternative abbreviations. I'll take it out if the tests still pass.